### PR TITLE
Encrypt each share under member's public key (cryptobox)

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves items 3., 4. and most of 7. on #51.

This PR introduces a wrapper around the cryptobox Hacl Nacl API to perform public-key encryption. This is used by CCF to encrypt the shares of `k_z` for each active member on startup and then record these encrypted shares in the `ccf.shares` public table. 

Since cryptobox provides authenticated encryption, whoever encrypts also needs to specify their public key. So the network now owns a public/private encryption key pair (`encryption_priv_key`) that is passed to other nodes via the join RPC response.

CCF also parses the members' encryption keys. Since these (x25519) keys are generated via openssl (see `keygenerator.sh`), we need to parse the asn1 public key in CCF (see `25519.h`). 

Next: 
- Adds `getEncryptedShare` RPC for members to retrieve their shares (5.)
- Let members decrypt their shares and submit it to CCF before recovery (6.)

### CODE REVIEW

File: src/tls/keypair.h [hunk 1]
```diff
@@ -172,29 +172,6 @@ namespace tls
       return Pem({data, len});
     }
 
-    /**
-     * Get the public key in ASN.1 format
-     */
-    std::vector<uint8_t> public_key_asn1()
```
Comment (jumaffre) [inline src/tls/keypair.h]: This was unused.

**Comment by cimetrics** [comment]
encrypt_shares@[5118 aka 20200220.8](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5118&view=results) vs master ewma over 30 builds from [4808](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4808&view=results) to [5114](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5114&view=results)
images

File: src/crypto/cryptobox.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  public:
+    static constexpr size_t KEY_SIZE = 32;
+
+    static std::vector<uint8_t> public_from_private(
+      std::vector<uint8_t>& private_key)
+    {
+      if (private_key.size() != KEY_SIZE)
+      {
+        throw std::logic_error(
+          fmt::format("Private key size is not {}", KEY_SIZE));
```
Comment (olgavrou) [inline src/crypto/cryptobox.h]: would it be helpful here in all the errors where the key size is not correct to also print the key size that we got?

File: src/crypto/cryptobox.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      if (
+        Hacl_NaCl_crypto_box_easy(
+          cipher.data(),
+          plain.data(),
+          plain.size(),
+          nonce.data(),
+          recipient_public.data(),
+          sender_private.data()) != 0)
+      {
+        throw std::logic_error("Box create(): encryption failed");
```
Comment (olgavrou) [inline src/crypto/cryptobox.h]: Add the failure code to the error message?

File: src/crypto/cryptobox.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      if (
+        Hacl_NaCl_crypto_box_open_easy(
+          plain.data(),
+          (uint8_t*)cipher.data(),
+          cipher.size(),
+          nonce.data(),
+          (uint8_t*)sender_public.data(),
+          (uint8_t*)recipient_private.data()) != 0)
+      {
+        throw std::logic_error("Box open(): decryption failed");
```
Comment (olgavrou) [inline src/crypto/cryptobox.h]: same here

File: src/tls/25519.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "pem.h"
+#include "tls.h"
+
+namespace tls
+{
```
Comment (olgavrou) [inline src/tls/25519.h]: are these tested? also would it provide additional information on the errors thrown to add the failure code?

File: src/tls/25519.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      mbedtls_pem_read_buffer(
+        &pem,
+        "-----BEGIN PUBLIC KEY-----",
+        "-----END PUBLIC KEY-----",
+        public_pem.data(),
+        nullptr,
+        0,
+        &pem_len) != 0)
+    {
+      throw std::logic_error("parse_25519_public(): Failed to read PEM");
```
Comment (eddyashton) [inline src/tls/25519.h]: We should include `tls::error_string(rc)` in these errors.

File: src/tls/25519.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#pragma once
+
+#include "pem.h"
+#include "tls.h"
+
+namespace tls
+{
+  // This function parses x25519 PEM keys generated by openssl (e.g. for
+  // members' public encryption key) and returns the raw 32-byte key.
+  static std::vector<uint8_t> parse_25519_public(const Pem& public_pem)
```
Comment (eddyashton) [inline src/tls/25519.h]: I'm surprised we can't do this with the existing `tls::PublicKey`, I think `mbedtls_pk_write_pubkey_der` would write these raw key bytes?

Comment (jumaffre) [reply src/tls/25519.h]: I thought so too. However, we'd need to load the PEM into a `mbedtls_pk_context` first and this does not work as `25519` keys are not support by the version of TLS we use.

File: src/crypto/cryptobox.h (hunk 3 above)
Comment (jumaffre) [reply src/crypto/cryptobox.h]: The Hacl Nacl API only returns 0 on success or -1 on error.

File: src/tls/25519.h (hunk 5 above)
Comment (jumaffre) [reply src/tls/25519.h]: Good point! I added a test for it and include the `tls::error_string(rc)` throughout the file.

File: src/tls/25519.h (hunk 7 above)
Comment (jumaffre) [reply src/tls/25519.h]: I have added a comment to mention this limitation at the top of the file.

Comment (eddyashton) [reply src/tls/25519.h]: Ah of course, fair enough!

**Pull request merged**